core/src/main/java/org/infinispan/interceptors/distribution/TxDistributionInterceptor.java

```diff
@@ -247,47 +252,66 @@ public Object visitRollbackCommand(TxInvocationContext ctx, RollbackCommand comm
       return localTx.getCommitNodes(affectedNodes, rpcManager.getTopologyId(), members);
    }
 
-   private void sendCommitCommand(TxInvocationContext ctx, CommitCommand command) throws TimeoutException, InterruptedException {
-      Collection<Address> recipients = getCommitNodes(ctx);
-      boolean syncCommitPhase = cacheConfiguration.transaction().syncCommitPhase();
-      RpcOptions rpcOptions;
-      if (syncCommitPhase) {
-         rpcOptions = rpcManager.getRpcOptionsBuilder(ResponseMode.SYNCHRONOUS_IGNORE_LEAVERS, DeliverOrder.NONE).build();
-      } else {
-         rpcOptions = rpcManager.getDefaultRpcOptions(false, DeliverOrder.NONE);
-      }
-      Map<Address, Response> responseMap = rpcManager.invokeRemotely(recipients, command, rpcOptions);
-      checkTxCommandResponses(responseMap, command);
-   }
-
-   protected void checkTxCommandResponses(Map<Address, Response> responseMap, TransactionBoundaryCommand command) {
+   protected void checkTxCommandResponses(Map<Address, Response> responseMap, TransactionBoundaryCommand command,
+                                          LocalTxInvocationContext context, Collection<Address> recipients) {
+      boolean handledByPartitionHandling = false;
+      boolean throwOutdatedTopologyException = false;
       for (Map.Entry<Address, Response> e : responseMap.entrySet()) {
          Address recipient = e.getKey();
          Response response = e.getValue();
-         if (response instanceof CacheNotFoundResponse) {
+         if (response == CacheNotFoundResponse.INSTANCE) {
             // No need to retry if the missing node wasn't a member when the command started.
             if (command.getTopologyId() == stateTransferManager.getCacheTopology().getTopologyId()
                   && !rpcManager.getMembers().contains(recipient)) {
-               log.tracef("Ignoring response from node not targeted %s", recipient);
+               if (trace) log.tracef("Ignoring response from node not targeted %s", recipient);
             } else {
-               log.tracef("Cache not running on node %s, or the node is missing", recipient);
-               throw new OutdatedTopologyException("Cache not running on node " + recipient);
+               if (!handledByPartitionHandling && checkResponseInPartitionHandling(command, context, recipients)) {
```

Comment:
This `if` looks too complicated to me. I'd either call `checkResponseInPartitionHandling` and log the same message every time, or return out of the function the moment we get the first partition handling error.